service/core/serverObj/socks.go

```diff
@@ -0,0 +1,133 @@
+package serverObj
+
+import (
+	"net"
+	"net/url"
+	"strconv"
+
+	"github.com/v2rayA/v2rayA/core/coreObj"
+)
+
+func init() {
+	FromLinkRegister("socks-proxy", NewSOCKS)
+	FromLinkRegister("socks", NewSOCKS)
+	EmptyRegister("socks-proxy", func() (ServerObj, error) {
+		return new(SOCKS), nil
+	})
+	EmptyRegister("socks", func() (ServerObj, error) {
+		return new(SOCKS), nil
+	})
+}
+
+type SOCKS struct {
+	Name     string `json:"name"`
+	Server   string `json:"server"`
+	Port     int    `json:"port"`
+	Username string `json:"username"`
+	Password string `json:"password"`
+	Protocol string `json:"protocol"`
+}
+
+func NewSOCKS(link string) (ServerObj, error) {
+	return ParseSocksURL(link)
+}
+
+func ParseSocksURL(u string) (data *SOCKS, err error) {
+	t, err := url.Parse(u)
+	if err != nil {
+		return nil, InvalidParameterErr
+	}
+	port, err := strconv.Atoi(t.Port())
+	if err != nil {
+		return nil, InvalidParameterErr
+	}
+	data = &SOCKS{
+		Name:   t.Fragment,
+		Server: t.Hostname(),
+		Port:   port,
+	}
+	if t.User != nil && len(t.User.String()) > 0 {
+		data.Username = t.User.Username()
+		data.Password, _ = t.User.Password()
+	}
+	switch t.Scheme {
+	case "socks", "socks-proxy":
+		data.Protocol = "socks"
+		if data.Port == 0 {
+			data.Port = 1080
+		}
+	default:
+		data.Protocol = t.Scheme
+	}
+	return data, nil
+}
+
+func (h *SOCKS) Configuration(info PriorInfo) (c Configuration, err error) {
+	var users []coreObj.OutboundUser
+	if h.Username != "" && h.Password != "" {
```

Comment:
What about non-empty `Username` with empty `Password`? I don't know if it is valid.

Reply:
> 

Hoo that was interesting. I tested on v2ray <-> v2ray and seems leaving either username or password empty are both valid. SOCKS5 authentication is quite rare though.

If we wish to support this, `util.js:153` might also need some change? It turns password into username. 

Any way, I don't think that's a good idea, hmm

Reply:
Yes. It does not matter. As what you say, it is rare to use.